DiverseSelector/base.py

```diff
@@ -24,71 +24,123 @@
+
 from .utils import feature_generator, feature_reader, get_features
 
+class Selector(ABC):
+    @abstractmethod
+    def pick_initial_compounds(self):
+        """Pick the initial compounds.
+        """
+        pass
+
+    @abstractmethod
+    def compute_diversity(self, metric: str, func=None):
+        """
+        Compute the diversity between every component in dataset
+        :param metric: one of the standard metrics ("Euclidean", "Cosine", "Jaccard" etc. )
+        :param func: callable function that take 2 arguments and calculate distance between them. Specified by user
+        :return: np.ndarray aka (dis-)similarity matrix
+        """
+        pass
+
+    @abstractmethod
+    def select(self, num_selected: int, random_seed: int = None):
+        """
+        Select the subset molecules with optimal diversity.
+        :param num_selected: amount of molecule that need to be selected
+        :param random_seed: random seed for random selection compounds
+        :return: np.array of indices of selected molecules from the initial dataset ## Maybe not indices....
+        """
+        pass
 
-class DissimilaritySelectionBase(ABC):
+    @abstractmethod
+    def subset_diversity(self, subset: np.ndarray, div_metric: str):
+        """
+        Calculate diversity of the subset.
+        :param subset: numpy array of indices which make up a subset
+        :param div_metric: metric for calculating diversity of the subset ("Gini", "Entropy" etc.)
+        :return: float #Scale should be discussed (0 to 1, or 0 to 100 or -1 to 1 or anything else)
+        """
+        # todo: need to implement diversity measurement here
+        pass
+
+    @abstractmethod
+    @property
+    def all_diversity(self, div_metric: str): # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Calculate diversity of the original dataset.
+        :param div_metric: metric for calculating diversity of the subset ("Gini", "Entropy" etc.)
+        :return: float #Scale should be discussed (0 to 1, or 0 to 100 or -1 to 1 or anything else)
+        """
+        # todo: need to implement diversity measurement here
+        pass
+
+    def read_excel(self, fname: str):  # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Read dataset from an excel file
+        :param fname: file name for saving molecule
+        :return: None
+        """
+        pass
+
+    def to_excel(self, fname: str):  # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Save selected molecules as an excel file
+        :param fname: file name for saving molecule
+        :return: None
+        """
+        pass
+
+    def to_txt(self, fname: str): # Not an abstract method!!! Needs to be implemented explicitly here
+        """
+        Save selected molecules as an excel file
+        :param fname: file name for saving molecule
+        :return: None
+        """
+        pass
```

Comment:
Why didn't we just build one abstract class for dissimilarity based compound selection?

Reply:
My idea is based on the fact that we have only _one_ abstract class called `Selector`. This abstract class should define behaviour of any other class selector that is based on family of methods: ClusterSelector, DissimilaritySelector, maybe even MonteCarlo selector etc. . In this way we avoid several issues:
1. Unnecessary creation of _two_ abstract classes that still have some common methods (that needs to be copying) 
2. Duplicating the same functionality again while creating _concrete_ class for DissimilaritySelector.
Just to sum up: I believe we will have only one `DissimilaritySelector` class that combine different metrics and algorithm. So I don't think we need to have an abstract class (which is used for describing behaviour for _several_ child classes) that will be inherited by only _one_ class

Reply:
Your comment makes sense. But then for `ClusterSelector`, we don't need `pick_initial_compounds` method. I don't know if we can come up with a good way of doing this. Therefore, I think having two abstract classes is a sensible choice. What do you think? @RichRick1

Reply:
I'm not sure what `pick_initial_compouds` method should do tbh. As long as it doesn't present in both classes we can just delete it from the abstract class and implement it only in DissimilaritySelector and still have the same structure